File: internal/transport/stdio.go
// https://hypercall.xyz

package transport

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"sync"

	"github.com/hypercall-public/mcpzip/internal/types"
	"github.com/modelcontextprotocol/go-sdk/mcp"
)

// StdioUpstream connects to an upstream MCP server via stdio using the go-sdk.
type StdioUpstream struct {
	mu      sync.Mutex
	session *mcp.ClientSession
	client  *mcp.Client
	alive   bool
}

// NewStdioUpstream creates and connects to an upstream stdio MCP server.
func NewStdioUpstream(ctx context.Context, name string, cfg types.ServerConfig) (*StdioUpstream, error) {
	client := mcp.NewClient(
		&mcp.Implementation{Name: "mcpzip", Version: "0.1.0"},
		nil,
	)

	cmd := exec.CommandContext(ctx, cfg.Command, cfg.Args...)
	if len(cfg.Env) > 0 {
		cmd.Env = os.Environ()
		for k, v := range cfg.Env {
			cmd.Env = append(cmd.Env, k+"="+v)
		}
	}

	session, err := client.Connect(ctx, &mcp.CommandTransport{Command: cmd}, nil)
	if err != nil {
		return nil, fmt.Errorf("connecting to %q: %w", name, err)
	}

	return &StdioUpstream{
		session: session,
		client:  client,
		alive:   true,
	}, nil
}

func (s *StdioUpstream) ListTools(ctx context.Context) ([]types.ToolEntry, error) {
	s.mu.Lock()
	session := s.session
	s.mu.Unlock()
	if session == nil {
		return nil, fmt.Errorf("connection closed")
	}

	var allTools []*mcp.Tool
	for tool, err := range session.Tools(ctx, nil) {
		if err != nil {
			return nil, fmt.Errorf("listing tools: %w", err)
		}
		allTools = append(allTools, tool)
	}

	entries := make([]types.ToolEntry, 0, len(allTools))
	for _, t := range allTools {
		schemaBytes, err := json.Marshal(t.InputSchema)
		if err != nil {
			schemaBytes = []byte(`{"type":"object"}`)
		}
		entries = append(entries, types.ToolEntry{
			OriginalName: t.Name,
			Description:  t.Description,
			InputSchema:  schemaBytes,
			CompactParams: types.CompactParamsFromSchema(schemaBytes),
		})
	}
	return entries, nil
}

func (s *StdioUpstream) CallTool(ctx context.Context, toolName string, args json.RawMessage) (json.RawMessage, error) {
	s.mu.Lock()
	session := s.session
	s.mu.Unlock()
	if session == nil {
		return nil, fmt.Errorf("connection closed")
	}

	// Convert json.RawMessage args to map[string]any for the SDK.
	var argsMap map[string]any
	if len(args) > 0 {
		if err := json.Unmarshal(args, &argsMap); err != nil {
			return nil, fmt.Errorf("unmarshaling tool arguments: %w", err)
		}
	}

	result, err := session.CallTool(ctx, &mcp.CallToolParams{
		Name:      toolName,
		Arguments: argsMap,
	})
	if err != nil {
		return nil, err
	}

	// Convert CallToolResult back to JSON.
	// If the result has structured content, use that.
	if result.StructuredContent != nil {
		return json.Marshal(result.StructuredContent)
	}

	// Otherwise, extract text content.
	if len(result.Content) == 1 {
		if tc, ok := result.Content[0].(*mcp.TextContent); ok {
			// Try to parse as JSON first; if it's already JSON, return as-is.
			if json.Valid([]byte(tc.Text)) {
				return json.RawMessage(tc.Text), nil
			}
			return json.Marshal(tc.Text)
		}
	}

	// Fallback: marshal the entire content array.
	return json.Marshal(result.Content)
}

func (s *StdioUpstream) Close() error {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.alive = false
	if s.session != nil {
		err := s.session.Close()
		s.session = nil
		return err
	}
	return nil
}

func (s *StdioUpstream) Alive() bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.alive && s.session != nil
}

File: src/transport/stdio.rs
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;

use async_trait::async_trait;
use serde_json::Value;
use tokio::process::Command;

use crate::error::McpzipError;
use crate::mcp::client::McpClient;
use crate::mcp::protocol::{ContentItem, ToolInfo};
use crate::mcp::transport::NdjsonTransport;
use crate::transport::Upstream;
use crate::types::{compact_params_from_schema, prefixed_name, ServerConfig, ToolEntry};

/// Upstream connection via a stdio subprocess.
pub struct StdioUpstream {
    name: String,
    client: McpClient,
    alive: AtomicBool,
    child: tokio::sync::Mutex<Option<tokio::process::Child>>,
}

impl StdioUpstream {
    /// Spawn a subprocess and perform MCP handshake.
    pub async fn new(name: String, cfg: &ServerConfig) -> Result<Self, McpzipError> {
        let command = cfg
            .command
            .as_deref()
            .ok_or_else(|| McpzipError::Config(format!("server {:?}: missing command", name)))?;

        let mut cmd = Command::new(command);
        if let Some(args) = &cfg.args {
            cmd.args(args);
        }
        if let Some(env) = &cfg.env {
            for (k, v) in env {
                cmd.env(k, v);
            }
        }

        cmd.stdin(std::process::Stdio::piped())
            .stdout(std::process::Stdio::piped())
            .stderr(std::process::Stdio::inherit());

        let mut child = cmd
            .spawn()
            .map_err(|e| McpzipError::Transport(format!("failed to spawn {:?}: {}", command, e)))?;

        let stdin = child
            .stdin
            .take()
            .ok_or_else(|| McpzipError::Transport("failed to capture stdin".into()))?;
        let stdout = child
            .stdout
            .take()
            .ok_or_else(|| McpzipError::Transport("failed to capture stdout".into()))?;

        let transport = Arc::new(NdjsonTransport::new(Box::new(stdout), Box::new(stdin)));
        let client = McpClient::new(transport);

        // Perform MCP handshake
        client.initialize().await.map_err(|e| {
            McpzipError::Transport(format!("handshake failed for {:?}: {}", name, e))
        })?;

        Ok(Self {
            name,
            client,
            alive: AtomicBool::new(true),
            child: tokio::sync::Mutex::new(Some(child)),
        })
    }
}

#[async_trait]
impl Upstream for StdioUpstream {
    async fn list_tools(&self) -> Result<Vec<ToolEntry>, McpzipError> {
        let tools = self.client.list_tools().await?;
        Ok(tools
            .into_iter()
            .map(|t| tool_info_to_entry(&self.name, t))
            .collect())
    }

    async fn call_tool(&self, tool_name: &str, args: Value) -> Result<Value, McpzipError> {
        let result = self.client.call_tool(tool_name, args).await?;

        // Convert CallToolResult to raw JSON value, matching Go behavior:
        // - Single text content that's valid JSON: return the text as raw JSON
        // - Otherwise: return the full result as JSON
        if result.content.len() == 1 {
            if let ContentItem::Text { ref text } = result.content[0] {
                if serde_json::from_str::<Value>(text).is_ok() {
                    return Ok(serde_json::from_str(text)?);
                }
                return Ok(Value::String(text.clone()));
            }
        }
        Ok(serde_json::to_value(&result)?)
    }

    async fn close(&self) -> Result<(), McpzipError> {
        self.alive.store(false, Ordering::Relaxed);
        let mut child_guard = self.child.lock().await;
        if let Some(mut child) = child_guard.take() {
            let _ = child.kill().await;
        }
        Ok(())
    }

    fn alive(&self) -> bool {
        self.alive.load(Ordering::Relaxed)
    }
}

fn tool_info_to_entry(server_name: &str, info: ToolInfo) -> ToolEntry {
    let schema = info.input_schema.unwrap_or(Value::Null);
    let compact = compact_params_from_schema(&schema);
    ToolEntry {
        name: prefixed_name(server_name, &info.name),
        server_name: server_name.into(),
        original_name: info.name,
        description: info.description.unwrap_or_default(),
        input_schema: schema,
        compact_params: compact,
    }
}
